google/cloud/storage/emulator/gcs/holder.py

```diff
@@ -59,21 +59,31 @@ def init_upload(
 
     @classmethod
     def init_resumable_rest(cls, request, bucket):
-        name = request.args.get("name", "")
+        name = request.args.get("name", None)
         rest_only = {}
-        if len(request.data) > 0 and request.data.decode("utf-8") != "{}":
-            if name != "":
-                utils.error.invalid("name argument in non-empty payload", None)
+        if len(request.data) > 0:
             data = json.loads(request.data)
             rest_only = cls.__extract_rest_only(data)
             metadata = json_format.ParseDict(data, resources_pb2.Object())
+            if name != None and name != metadata.name:
+                utils.error.invalid(
+                    "Value '%s' in content does not agree with value '%s'."
+                    % (metadata.name, name),
+                    context=None,
+                )
+            if not metadata.name and name:
+                metadata.name = name
         else:
             metadata = resources_pb2.Object()
+            if not metadata.name:
+                utils.error.invalid("No object name", context=None)
             metadata.name = name
         if metadata.content_type == "":
             metadata.content_type = request.headers.get(
                 "x-upload-content-type", "application/octet-stream"
             )
+        if metadata.name == "":
+            utils.error.invalid("No object name", context=None)
```

Comment:
It might be useful to use a different error message, if only to troubleshoot the CI failures.